client/app/i18n/index.js

```diff
@@ -0,0 +1,22 @@
+import i18n from "i18next";
+import LanguageDetector from "i18next-browser-languagedetector";
+import { initReactI18next } from "react-i18next";
+import en from "./locales/en.json";
+import zh from "./locales/zh.json";
```

Comment:
Out of curiosity, with this `zh` I've often seen it split into two.  One for "Simplified Chinese", and one for "Traditional Chinese".

Should we have this named as one of them, to leave room for the other one at some future point?

Reply:
> Out of curiosity, with this `zh` I've often seen it split into two. One for "Simplified Chinese", and one for "Traditional Chinese".
> 
> Should we have this named as one of them, to leave room for the other one at some future point?

OK. You are right.

Reply:
One note is that importing languages like this will add all of them into the bundle, which may not be nice 😅.

It's possibly better to do an async import for those cases. https://www.i18next.com/how-to/add-or-load-translations had a bit of content on this.

---

Separate thing I was thinking about these days: Should we plan ahead how we will do the translations end-to-end, before committing anything? I don't know much about `react-i18next`, I've used mostly [react-intl](https://www.npmjs.com/package/react-intl), the main difference I saw so far is that the later also has a `description` for each unformatted string. Considering Redash as an open-source / community effort and how LLMs are today, it may be a good thing to plan a flow which includes automatic generation of the translation files on every release. Then would count on the community only for reviewing/fixing eventual mistakes.

Reply:
> One note is that importing languages like this will add all of them into the bundle, which may not be nice 😅.
> 
> It's possibly better to do an async import for those cases. https://www.i18next.com/how-to/add-or-load-translations had a bit of content on this.
> 
> Separate thing I was thinking about these days: Should we plan ahead how we will do the translations end-to-end, before committing anything? I don't know much about `react-i18next`, I've used mostly [react-intl](https://www.npmjs.com/package/react-intl), the main difference I saw so far is that the later also has a `description` for each unformatted string. Considering Redash as an open-source / community effort and how LLMs are today, it may be a good thing to plan a flow which includes automatic generation of the translation files on every release. Then would count on the community only for reviewing/fixing eventual mistakes, in this case the `description` would be useful for providing context and/or fixing the prompt.

Cool. It's a good idea.

Reply:
> One note is that importing languages like this will add all of them into the bundle, which may not be nice 😅.
> 
> It's possibly better to do an async import for those cases. https://www.i18next.com/how-to/add-or-load-translations had a bit of content on this.
> 
> Separate thing I was thinking about these days: Should we plan ahead how we will do the translations end-to-end, before committing anything? I don't know much about `react-i18next`, I've used mostly [react-intl](https://www.npmjs.com/package/react-intl), the main difference I saw so far is that the later also has a `description` for each unformatted string. Considering Redash as an open-source / community effort and how LLMs are today, it may be a good thing to plan a flow which includes automatic generation of the translation files on every release. Then would count on the community only for reviewing/fixing eventual mistakes, in this case the `description` would be useful for providing context and/or fixing the prompt.

I haven't figured out which platform to host these mapping tables for the time being.

Reply:
As a data point - for a project many years ago - we had the project hosted on Launchpad, so it could make use of the translation infrastructure there.

One *really* cool benefit of that platform at the time was that if anyone, anywhere had translated the same string for any project on Launchpad (with compatible license) then the string was made available as a potential option.

It saved a *lot* of time for people, and made it easy to get pretty far with new language translations.

That being said, the company behind Launchpad (Canonical) have done some shady stuff in the past so I'm not really a fan of theirs... :wink:

Reply:
The possible use of LLM's for generating "first attempt" translations sounds like a useful idea too.

Reply:
> The possible use of LLM's for generating "first attempt" translations sounds like a useful idea too.

Cool!